geometric/molecule.py

```diff
@@ -1687,6 +1702,74 @@ def radius_of_gyration(self):
+            for j, xj in enumerate(dxyz):
+                factor = PeriodicTable[self.elem[j]] if mass else 1.0
+                I += factor*(np.dot(xj,xj)*np.eye(3) - np.outer(xj,xj))
+            moments.append(I)
+        return moments
+        
+    def calc_netforce_torque(self, mass=True):
+        """ Calculate net force and torque vectors
+        in units of hartree/bohr and hartree/bohr*bohr respectively.
+
+        These are actually the "negative" of the net force and torque
+        because the factor of -1 to convert grad into force has not been applied.
+
+        Requires forces to be entered into qm_grads.
+        """
+        netforces = []
+        torques = []
+        coms = self.center_of_mass(mass=mass)
+        if len(self.qm_grads) != len(self):
+            raise RuntimeError('qm_grads length does not match number of structures')
+        if self.qm_grads[0].shape != (self.na, 3):
+            raise RuntimeError('qm_grads element does not have the wrong shape (n_atoms, 3)')
+        for i in range(len(self)):
+            netforces.append(np.sum(self.qm_grads[i], axis=0))
+            torque_vec = np.zeros(3, dtype=float)
+            dxyz = (self.xyzs[i] - coms[i][np.newaxis, :])/bohr2ang
+            for j in range(self.na):
+                torque_vec += np.cross(dxyz[j], self.qm_grads[i][j])
+            torques.append(torque_vec)
+        return np.array(netforces), np.array(torques)
+
+    def remove_netforce_torque(self, mass=True):
+        """ Calculate net force and torque-less analogue of qm_grads. """
+        netforces, torques = self.calc_netforce_torque(mass=mass)
+        coms = self.center_of_mass(mass=mass)
+        moments = self.moment_of_inertia(mass=mass)
+        grad_proj = [frc.copy() for frc in self.qm_grads]
+        for i in range(len(self)):
+            # The purely translational component of the force on each atom
+            trans_frc = netforces[i] / self.na
+            grad_proj[i] -= trans_frc[np.newaxis, :]
+            dxyz = (self.xyzs[i] - coms[i][np.newaxis, :])/bohr2ang
+            # Moment of inertia in amu bohr**2
+            moment_au = moments[i]/bohr2ang**2
+            for j in range(self.na):
+                # The purely rotational component of the force on each atom.
+                # L = angular momentum, I = moment of inertia, w = angular velocity
+                # T = torque, v = perpendicular component of velocity
+                # F = perpendicular component of force, the desired quantity
+                #
+                # Start with F = m (dv/dt) = m (dw/dt x r)
+                # dw/dt = I^-1.dL/dt = I^-1.T
+                # Therefore F = m (I^-1.T) x r
+                I_torque = np.dot(np.linalg.pinv(moment_au), torques[i])
+                factor = PeriodicTable[self.elem[j]] if mass else 1.0
+                torque_frc = factor*np.cross(I_torque, dxyz[j])
+                grad_proj[i][j] -= torque_frc
+        return grad_proj
```

Comment:
I checked the net force and torque removal code and it looks perfect! As formulated, the total molecular force on the center of mass and the total molecular torque wrt the center of mass are subtracted from the energy gradient.

Reply:
Thank you very much for looking into this!  I think it will improve the performance of constrained optimizations in Psi4. :)